datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -275,68 +367,93 @@ impl CommonSubexprEliminate {
         config: &dyn OptimizerConfig,
     ) -> Result<Transformed<LogicalPlan>> {
         // collect all window expressions from any number of LogicalPlanWindow
-        let ConsecutiveWindowExprs {
-            window_exprs,
-            arrays_per_window,
-            expr_stats,
-            plan,
-        } = ConsecutiveWindowExprs::try_new(window)?;
+        let (mut window_exprs, mut window_schemas, mut plan) =
+            get_consecutive_window_exprs(window);
 
-        let arrays_per_window = arrays_per_window
+        let mut found_common = false;
+        let mut expr_stats = ExprStats::new();
+        let arrays_per_window = window_exprs
             .iter()
-            .map(|arrays| arrays.as_slice())
-            .collect::<Vec<_>>();
+            .map(|window_expr| {
+                self.to_arrays(window_expr, &mut expr_stats, ExprMask::Normal)
+                    .map(|(fc, id_arrays)| {
+                        found_common |= fc;
 
-        // save the original names
-        let name_preserver = NamePreserver::new(&plan);
-        let mut saved_names = window_exprs
-            .iter()
-            .map(|exprs| {
-                exprs
-                    .iter()
-                    .map(|expr| name_preserver.save(expr))
-                    .collect::<Result<Vec<_>>>()
+                        id_arrays
+                    })
             })
             .collect::<Result<Vec<_>>>()?;
 
-        assert_eq!(window_exprs.len(), arrays_per_window.len());
-        let num_window_exprs = window_exprs.len();
-        let rewritten_window_exprs = self.rewrite_expr(
-            window_exprs,
-            &arrays_per_window,
-            plan,
-            &expr_stats,
-            config,
-        )?;
-        let transformed = rewritten_window_exprs.transformed;
+        if found_common {
+            // save the original names
+            let name_preserver = NamePreserver::new(&plan);
+            let mut saved_names = window_exprs
+                .iter()
+                .map(|exprs| {
+                    exprs
+                        .iter()
+                        .map(|expr| name_preserver.save(expr))
+                        .collect::<Result<Vec<_>>>()
+                })
+                .collect::<Result<Vec<_>>>()?;
 
-        let (mut new_expr, new_input) = rewritten_window_exprs.data;
+            assert_eq!(window_exprs.len(), arrays_per_window.len());
+            let num_window_exprs = window_exprs.len();
+            let rewritten_window_exprs = self.rewrite_expr(
+                window_exprs.clone(),
+                arrays_per_window,
+                plan,
+                &expr_stats,
+                config,
+            )?;
+            let transformed = rewritten_window_exprs.transformed;
+            assert!(transformed);
```

Comment:
Now that `to_arrays()` return flags to indicate wether it make seense to run the 2nd, replacing traversal it is not necessary to return `Transformed` in `rewrite_expr()` / `rewrite_exprs_list()` / `replace_common_expr()`, but it does no harm and we can assert that the rewritten's `.transformed` is true.